File: tests/ut/cpp/ops/test_ops_searchsorted.cc
/**
 * Copyright 2024 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include <memory>
#include "common/common_test.h"
#include "ir/dtype/type.h"
#include "abstract/dshape.h"
#include "utils/tensor_construct_utils.h"
#include "ir/primitive.h"
#include "abstract/abstract_value.h"
#include "include/backend/optimizer/helper.h"
#include "ops/test_ops.h"
#include "ops/ops_func_impl/searchsorted.h"
#include "ops/test_value_utils.h"

namespace mindspore {
namespace ops {

struct SearchSortedParam {
  std::vector<int64_t> sequence_shape;
  std::vector<int64_t> values_shape;
  std::vector<int64_t> sorter_shape;
  TypePtr in_dtype;
  ValuePtr out_dtype;
  bool right;
};

class TestSearchSorted : public TestOps, public testing::WithParamInterface<SearchSortedParam> {};

TEST_P(TestSearchSorted, SearchSorted_dyn_shape) {
  const auto &param = GetParam();

  SearchSortedFuncImpl func_impl;
  auto prim = std::make_shared<Primitive>("SearchSorted");
  auto sequence = std::make_shared<abstract::AbstractTensor>(param.in_dtype, param.sequence_shape);
  auto values = std::make_shared<abstract::AbstractTensor>(param.in_dtype, param.values_shape);
  auto sorter = std::make_shared<abstract::AbstractTensor>(kInt64, param.sorter_shape);
  auto right = CreateScalar(param.right)->ToAbstract();
  auto dtype = param.out_dtype->ToAbstract();

  auto expect_shape = std::make_shared<abstract::TensorShape>(param.values_shape);
  auto expect_dtype =
    std::make_shared<TensorType>(TypeIdToType(static_cast<TypeId>(param.out_dtype->cast<Int64ImmPtr>()->value())));

  auto out_shape = func_impl.InferShape(prim, {sequence, values, sorter, dtype, right});
  ASSERT_TRUE(*out_shape == *expect_shape);
  auto out_dtype = func_impl.InferType(prim, {sequence, values, sorter, dtype, right});
  ASSERT_TRUE(*out_dtype == *expect_dtype);
}

auto SearchSortedParamOpTestCases = testing::Values(
  SearchSortedParam{{10}, {10}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{10, 1, 2}, {10, 1, 10}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{10, 4, 2}, {10, 4, 10}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), false},
  SearchSortedParam{{10, 1, -1}, {10, 1, 10}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},

  SearchSortedParam{{10}, {10}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{10}, {10, 1, 2}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{10}, {10, 4, 2}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), false},
  SearchSortedParam{{10}, {10, 1, -1}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{10}, {-2}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},

  SearchSortedParam{{4, 5}, {4, 15}, {4, 5}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{
    {2, 1, 4, 5, 6, 9}, {2, 1, 4, 5, 6, 9}, {2, 1, 4, 5, 6, 9}, kFloat32, CreatePyInt(kNumberTypeInt32), false},
  SearchSortedParam{{2, 3, 4, -1}, {2, 3, 4, 5}, {2, 3, 4, -1}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{2, 3, 4, -1}, {-1, -1, 4, 5}, {2, 3, 4, -1}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{2, 1, 4, -1}, {-1, -1, 4, 5}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), false},
  SearchSortedParam{{2, 1, 4, 5, 6, 9}, {-2}, {2, 1, 4, 5, 6, 9}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{2, 1, 4, 5, -1, 9}, {-2}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), true},
  SearchSortedParam{{-2}, {-2}, {}, kFloat32, CreatePyInt(kNumberTypeInt32), false});

INSTANTIATE_TEST_CASE_P(TestSearchSorted, TestSearchSorted, SearchSortedParamOpTestCases);
}  // namespace ops
}  // namespace mindspore

File: tests/st/ops/test_ops_searchsorted.py
# Copyright 2023 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import numpy as np
import pytest
from tests.st.utils import test_utils
from tests.st.ops.dynamic_shape.test_op_utils import TEST_OP
from mindspore import ops, nn
import mindspore as ms

class SearchSortedNet(nn.Cell):
    def __init__(self):
        super(SearchSortedNet, self).__init__()
        self.searchsorted = ops.searchsorted

    def construct(self, x1, x2, out_int32=False, right=False, side="left", sorter=None):
        return self.searchsorted(x1, x2, out_int32=out_int32, right=right, side=side, sorter=sorter)


def generate_random_input(shape, dtype):
    return np.sort(np.random.uniform(0.9, 1.0, size=shape).astype(dtype)), \
           np.random.uniform(0.9, 1.0, size=shape).astype(dtype)

def searchsorted_forward_func(x, y, out_int32=False, right=False, sorter=None):
    return SearchSortedNet()(x, y, out_int32=out_int32, right=right, side="right", sorter=sorter)

def generate_expect_forward_output(x, y, side="right", sorter=None):
    return np.searchsorted(x, y, side=side, sorter=sorter)

@test_utils.run_with_cell
def searchsorted_vmap_func(x, y, out_int32=False, right=False, sorter=None):
    return ops.vmap(searchsorted_forward_func, in_axes=(0, 0, None, None, None),
                    out_axes=0)(x, y, out_int32, right, sorter)

@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
@pytest.mark.parametrize("context_mode", [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@test_utils.run_test_with_On
def test_searchsorted_forward(context_mode):
    """
    Feature: Ops.
    Description: test op searchsorted forward.
    Expectation: expect correct result.
    """
    ms.context.set_context(mode=context_mode)
    x_np = np.array([-1, 0.0, 1.0, 2.0], dtype=np.float32)
    y_np = np.array([0.0, 1.0, 2.0, -2], dtype=np.float32)
    x = ms.Tensor(x_np)
    y = ms.Tensor(y_np)
    out = searchsorted_forward_func(x, y)
    expect_out = generate_expect_forward_output(x_np, y_np)
    np.testing.assert_allclose(out.asnumpy(), expect_out, rtol=1e-3)

@pytest.mark.level1
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_ops_searchsorted_vmap(context_mode):
    """
    Feature: pyboost function.
    Description: test function searchsorted vmap.
    Expectation: expect correct result.
    """

    ms.context.set_context(mode=context_mode)
    x_np = np.array([[-1, 0.0, 1.0, 2.0], [-1, 0.0, 1.0, 2.0]], dtype=np.float32)
    y_np = np.array([[0.0, 1.0, 2.0, -2], [0.0, 1.0, 2.0, -2]], dtype=np.float32)
    x = ms.Tensor(x_np)
    y = ms.Tensor(y_np)
    out = searchsorted_vmap_func(x, y)
    expect_out = []
    for s_x_np, s_y_np in zip(x_np, y_np):
        np_out = generate_expect_forward_output(s_x_np, s_y_np)
        expect_out.append(np_out)

    np.testing.assert_allclose(out.asnumpy(), expect_out, rtol=1e-3)

@pytest.mark.level1
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_ops_searchsorted_dynamic(context_mode):
    """
    Feature: pyboost function.
    Description: test function searchsorted dynamic.
    Expectation: expect correct result.
    """

    x1, _ = generate_random_input((20), np.float32)
    _, other1 = generate_random_input((2, 3, 4, 5), np.float32)
    out_int32_1 = False
    right_1 = False
    sorter_1 = ms.Tensor(np.argsort(x1, axis=-1))

    x2, _ = generate_random_input((2, 3, 4), np.float32)
    _, other2 = generate_random_input((2, 3, 4), np.float32)
    out_int32_2 = True
    right_2 = True
    sorter_2 = ms.Tensor(np.argsort(x2, axis=-1))

    TEST_OP(searchsorted_forward_func,
            [[ms.Tensor(x1), ms.Tensor(other1), out_int32_1, right_1, sorter_1],
             [ms.Tensor(x2), ms.Tensor(other2), out_int32_2, right_2, sorter_2]], 'searchsorted')
